sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala

```diff
@@ -3902,6 +3902,15 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val RESPECT_USER_SCHEMA_NULLABILITY_FOR_FILE_DATA_SOURCES =
+    buildConf("spark.sql.respectUserSchemaNullabilityForFileDataSources")
+      .internal()
+      .doc("When true, the nullability in the user-specified schema for " +
+        "`DataFrameReader.schema(schema).json(path)` and .csv(path) and .xml(path) is respected" +
+        "Otherwise, they are turned to a nullable schema forcibly.")
+      .version("4.0.0")
+      .fallbackConf(LEGACY_RESPECT_NULLABILITY_IN_TEXT_DATASET_CONVERSION)
```

Comment:
why fall back to the config of text data source?

Reply:
This is some kind of follow up to PR introduced `LEGACY_RESPECT_NULLABILITY_IN_TEXT_DATASET_CONVERSION`, schema nullability is not preserved when path to file is given, even if flag is set to true. We had users who complained about not preserving schema nullability even if they set this flag. So to make them easier, fallback is set to the flag from previous PR, but since it is dangerous to rely on existing flag (because mitigation for potential issue will affect other code paths), I introduced new flag with fallback to the old one. So if something goes wrong, new flag can be set, while, if customer already overrode old flag to true, then there is no additional actions.